src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs

```diff
@@ -196,6 +197,27 @@ public virtual void ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags an
         {
             throw new NotImplementedException();
         }
+
+        /// <summary>
+        /// Attempts to compute or get the cached value provided by the given <paramref name="valueProvider"/> for the given <paramref name="key"/>.
+        /// Reusing the same <paramref name="valueProvider"/> instance across analyzer actions and/or analyzer instances can improve the overall analyzer performance by avoiding recomputation of the values.
+        /// </summary>
+        /// <typeparam name="TValue">The type of the value associated with the key.</typeparam>
+        /// <param name="key"><see cref="SourceText"/> for which the value is queried.</param>
+        /// <param name="valueProvider">Provider that computes the underlying value associated with the key.</param>
+        /// <param name="value">Value associated with the key.</param>
+        /// <returns>Returns true on success, false otherwise.</returns>
+        public bool TryGetValue<TValue>(SourceText key, SourceTextValueProvider<TValue> valueProvider, out TValue value)
```

Comment:
Well, even I thought about it, but felt that `SourceText` is indeed the key into the `SourceValueProvider` database from the client's perspective. A `TryGetValue` API with no explicit key parameter would seem odd as well.

Another alternative API design would be:

1. Define `RegisterXXXValueProvider` APIs on the context objects, which just takes a single value provider argument.
2. Define `TryGetValue(key, out TValue)` API on the `XXXValueProvider`